GRADLE-2123: remove dependencies from wtp component which are already in classpath (eclipse-wtp)

aaschmid opened the pull request:
I have implemented the behavior as discussed and confirmed with @adammurdoch in https://issues.gradle.org/browse/GRADLE-2123. Please also have a look at my last comment about the implementation in the issue!
The only thing which is still open/missing is if 'ear' is used together with 'java' or even 'war' plugin. I am not sure about it, yet. Any suggestions?

As @adammurdoch has already started changing some parts of the 'eclipse-wtp' plugin, I am looking forward to a discussion :-)
I would be glad to help more, however I think to get a clear implementation the 'classpath' and 'wtp component' has to know about each others dependencies to not even add incorrect values!

If you have any questions, don't hesitate to ask.

ajoberstar commented on subprojects/ide/src/main/groovy/org/gradle/plugins/ide/eclipse/EclipseWtpPlugin.groovy:
```diff
@@ -67,53 +69,60 @@ class EclipseWtpPlugin extends IdePlugin {
         project.plugins.withType(JavaPlugin) {
-            def deleteWtpDependentModule = []
             eclipseModel.classpath.file.whenMerged { Classpath classpath ->
                 if (hasWarOrEarPlugin(project)) {
```
Since WAR and Java (w/o WAR or EAR) are being handled identically now, besides the addition of the web libs container, it seems this check should just be for the EAR plugin.

aaschmid replied:
but what should happen if WAR and EAR plugin are applied (or what woudl you expect)? Currently, it would change the classpath and wtpComponent according to the behavior I would expect with WAR, after the changes it would change nothing if EAR is applied ...
What do you think?

ajoberstar replied:
Are you asking what behavior I would expect if both the EAR and WAR plugin are applied to a single project? If so, I've never had that use case, so I'm not sure.

aaschmid replied:
Yepp, that was the question. Due to easier code, I could push the small refactoring but as far the behavior ... I have no clue ...

ajoberstar replied:
Personally, I can't imagine why you would apply both, but it's probably an easier sell not to change the behavior.

mark-vieira replied:
Just a heads up. I've seen a few instances of this. Basically you create a single Gradle project which produces an EAR which contains a single WAR. Applying both the War and Ear plugin can satisfy this use case. You can think of it effectively as a WAR project, that **also** produces an EAR.

aaschmid replied:
Hi @mark-vieira, thanks for the Information. Does this mean the default settings of a WAR+EAR project should be from the WAR project and the EAR stuff has to be setup manually? Do you have an example project or build.gradle for me to write an automatic test for this?

mark-vieira replied:
An example would look something like [this](https://gist.github.com/mark-vieira/cfe83379e3b8749afc78).

aaschmid replied:
Hi @mark-vieira, just to be sure. For a project containing both war and ear plugin, where should the default settings come from?

mark-vieira replied:
To be honest, I'm not sure how we would consider modeling this scenario in Eclipse. Can you apply both facets to a project in Eclipse? I'm leaning towards treating it just like a normal EAR project, since it could, in theory, contain WARs from other traditional WAR subprojects.

ajoberstar commented on subprojects/ide/src/main/groovy/org/gradle/plugins/ide/eclipse/EclipseWtpPlugin.groovy:
```diff
@@ -67,53 +69,60 @@ class EclipseWtpPlugin extends IdePlugin {
-                        entry.entryAttributes[AbstractClasspathEntry.COMPONENT_NON_DEPENDENCY_ATTRIBUTE] = ''
-                    }
-                }
+            project.eclipse.classpath.file.whenMerged { Classpath classpath ->
```
If my comment above is addressed, these additions would be removed. The container would be the only piece addressed for the WAR plugin.

ajoberstar commented on subprojects/ide/src/main/groovy/org/gradle/plugins/ide/eclipse/EclipseWtpPlugin.groovy:
```diff
@@ -67,53 +69,60 @@ class EclipseWtpPlugin extends IdePlugin {
+                eclipseModel.classpath.fileReferenceFactory.fromFile(dependency.file).path
+            }
+            wtpComponent.wbModuleEntries.removeAll{ wbModule ->
+                wbModule instanceof WbDependentModule && depPaths.any{ path -> wbModule.handle.endsWith(path) }
```
It would be nice to avoid adding these in the first place, rather than removing them after the fact. That said, this should still get the job done.

aaschmid replied:
I know what you mean, but avoid adding all this completely needs a rather complete rewrite of the eclipse plugins behavior such that first retrieving all the information of all applied plugins bevor create all file considering all information together ...
If anybody of the Gradleware folks would agree on that and have a design idea, I would be glad to completely rewrite this plugins ...

ajoberstar commented:
I added some comments to the commits. Overall, this should do the job and it would be great to have in 2.4. Thanks for the work on these eclipse-wtp issues Andreas!

aaschmid commented:
Hi Gradle-Team,

I have rebased my changes that they can be merged ... 
I am wondering why this does neither get merged nor gets discussed. I think this is a great pity for contributers and I have the feeling of the time spent resolving this was for nothing :-(

Do you have at least a short comment on this?

Cheers,
Andreas

alkemist commented:
Hi Andreas,

This is all set to be merged, and we'll do so after we've branched for 2.4, so these changes will be in 2.5.

Thanks again.

aaschmid commented:
ok, thanks for the answer and that it will get merged :-)
(even if it's a pity that it's not scheduled for 2.4 as it really solves some IMHO bigger issues with the ```eclipse-wtp``` plugin)

aaschmid commented:
Hi @alkemist, 
I have resolved the merge conflicts and adjusted the behavior due to the changes in c3bb8f0 of @breskeby (see commit comment of d7e59e6 for details) so that you can review and hopefully merge it into master that we will get it released with v2.5.
Cheers, Andreas

alkemist commented:
Hi @aaschmid, we've missed the 2.5 release window unfortunately. This is high on our list for 2.6 and will be merged when one of our team members who is going to merge is back from holidays. Thanks for your patience.

aaschmid commented:
is there any progress on this?

aaschmid commented:
Will this fix ever be merged? Or is there any problem with it?

aaschmid commented:
Still no update @alkemist @adammurdoch ... keeping this open without any response gets more and more frustrating ...

ajoberstar commented:
I'd also like to reiterate my desire to have this improvement released.

bigdaz commented:
We appreciate the work put into this pull request.

Recently, the Gradle team hasn't had the capacity to deal with all of the bugs, feature requests, great ideas and pull requests that we receive.

Pull requests are interesting because they may seem like they come for "free". Our experience has shown that in order to maintain the level of quality and consistency that Gradle users expect, we need to invest significant engineering effort in all but the most trivial PRs.

However, we've been actively hiring, and I expect that you'll see progress on processing pull requests in the coming months. We do have a pretty big backlog, so it might take some time for us to take a look at this particular PR, particularly as Eclipse-WTP isn't something we are all familiar with.

Thanks for your patience.

aaschmid commented:
@bigdaz thanks for your long answer. I can really understand your points and that's why I tried to write real acceptance tests for this issue containing all the changes such that you can decide on the benefits. Please let me know if I can write more tests to satisfy anybody about the benefits of this not only bugfixing pull requests which also should bring more consistency to eclipse-wtp.

Cheers, Andreas

bigdaz commented:
Hi Andreas. We've recently hired some new engineers which should help us with the (daunting) task of tackling our PR backlog. I know that @pioterj has already started reviewing and prioritizing: I'll be sure to let him know that you are keen to work with us to get this merged.
Thanks again for your patience.
Daz

oehme commented:
Thank you for all your patience @aaschmid! I just had a closer look at this PR and there are a few issues we need to resolve:

`ClasspathFactory` has compile errors (or rather runtime errors, since it's Groovy). This was easy to resolve. 

After that the test `EclipseWtpModelIntegrationTest#wtp component respects configuration modifications` failed due to outdated expectations.

Another problem can be uncovered by slightly changing `EclipseWtpWebAndJavaProjectIntegrationTest`. Alter it so that the `:web` module uses guava version 19 and the `:java` module uses version 18. Guava 19 will be correctly added to the web classpath. But Guava 18 will be incorrectly added to the web component file.

There is another conceptual problem, probably because this change was made several Gradle versions ago: We no longer export classpath entries. So when `eclipse-wtp` is applied to a simple `java` project, **all** dependencies should be marked as `org.eclipse.jst.component.nondependency`. Otherwise you'll get the usual warning that it is neither exported nor deployed.

aaschmid commented:
Hi @oehme, unfortunately it is a long time ago since I have opened this pull request. With this short look I can just imagine the problems mentioned above. Could it be that Gradles way to handle Eclipse dependencies is purely wrong? Because if I remember correctly, there should be no ```org.eclipse.jst.component.nondependency``` by default, at least if you create and configure your project directly in Eclipse. 

Unfortunately, in the meantime I spent no time to follow the Gradle release notes and I am sure about the details of this issue ... What would be the expected behavior by using two different version of the same artifact? What is the behavior now?

oehme commented:
If you have a java+wtp project (a 'utility' project for your webapp) and you add a non-exported dependency, then you will get a warning saying `Classpathentry will not be exported or published. ClassNotFoundExceptions may result`. Adding the `org.eclipse.jst.component.nondependency` attribute fixes this.

Regarding the different versions: No external dependencies should be put in the *component* file. Gradle should do dependency resolution and then put the correct version (in this case 19) on the *classpath*.

aaschmid commented:
Ok, first thing still sounds strange to me but I think I have to look at it. 

For the second problem: I agree that no external dependencies should go into the component files and this is what I solved by this pull request! Is there a built-in functionality which doing the dependency comparison to tell me that DepA is just a different version of DepB and which one the newer is (sorry I cannot remember at the moment). I tried to use the internal mechanisms of Gradle to resolve dependencies. Did I do wrong?

oehme commented:
The unwanted dependencies are added in `WtpComponentFactory:91-98`. There, the factory transitively looks at the wtp models of upstream projects and adds their plus/minus configurations too. This results in the duplicate jar entries, because dependency resolution only works in the scope of a single project.

I think this code should be removed completely. Every project should be responsible of configuring its own plus/minus configurations correctly.

aaschmid commented:
Oh, now I got it. I changed the call using ```transitive := false``` in ```EclipseWtpPlugin``` and added a test case.

I also fixed the runtime errors after rebasing this branch onto the newest master of the original Gradle repository ...

I am just not sure what to do with ```org.gradle.plugins.ide.eclipse.EclipseWtpModelIntegrationTest#wtp component respects configuration modifications``` because IMHO it's obsolete. What do you think?

> In the same test you can also see that the :java project is added to the classpath of the :web project. 
> It should only be added to the component file instead. Otherwise you'll get a warning that it is neither 
> exported nor published.

Are you sure? How should the project compile without an classpath reference to the utility project?

> There is another conceptual problem, probably because this change was made several Gradle versions ago: We no longer export classpath entries. So when eclipse-wtp is applied to a simple java project, all dependencies should be marked as org.eclipse.jst.component.nondependency. Otherwise you'll get the usual warning that it is marked for publish, but not exported.

IMHO this is the case as if this is normal ```java``` project, neither ```plusConfigurations``` nor ```minusConfigurations``` should be specified such that for every classpath entry the line ```entry.entryAttributes[AbstractClasspathEntry.COMPONENT_NON_DEPENDENCY_ATTRIBUTE] = ''``` is executed.

Sorry, if my comments seemed a bit too rude above but after not looking at this for almost a year I forgot a lot about it and in the meanwhile I also got a bit pissed because it took some days to figure this out and it felt like being for the waste bin!

aaschmid commented:
I oversaw two test failures, let me quickly fix them as well.

aaschmid commented:
Please not that I updated the comment above :-)

oehme commented:
Hi Andreas,

first of all - no hard feelings, I can understand your frustration. We simply did not have the capacity to look into it, even though this change is very valuable.

Eclipse does not do proper dependency resolution. It just concatenates a project's classpath with the upstream project's exported entries. We want Gradle to calculate the correct classpath and tightly control it. That's why we no longer export any classpath entries in Eclipse.

Regarding the project dependency: The project is already referenced from the component file and thus it is contained in the "Web App Libraries" container that we add to the classpath. No need to add it a second time. You could also mark it as a 'nondependency' in the classpath to get rid of the warning. Or we'd have to get rid of the "Web App Libraries" container.

I took the example project attached to the Jira issue and the only way to get it warning free was:

###Java-Project

- all libraries marked as 'nondependency'

###Web-Project
- deployed libraries marked as 'dependency'
- non-deployed libraries marked as 'nondependency'
- project dependencies either removed from .classpath or marked as 'nondependency'

I think it would be best of you also had a look at that simple example to make sure we are on the same page :) In the meantime, I'll check out your current branch and look at the failing test.

oehme commented:
The failing test is due to another usage of the 'transitive' flag. This flag really stands in contrast to proper dependency resolution. I'll have to create some more elaborate examples to find out what this transitive search is trying to solve.

aaschmid commented:
Ok, the second call with transitive is done by ```WtpComponentFactory.groovy``` line 38 which is not touched by me. Unfortunately the commit 403ed271 which changed that does not tell me much about the reason for that while just having a very quick look. Changing it to ```false``` result in a failing test:

```
org.gradle.plugins.ide.eclipse.EclipseIntegrationTest > canCreateAndDeleteMetaData FAILED
```
But this was obvious. The question to me is still if the dependencies in Eclipse are correct if you export a war file from it (as easy test for starting an application server from Eclipse directly).

oehme commented:
Yes, that is what I'm trying to verify using examples. I hand-configured a project in the way I described above and it exported the correct war. Having some real world examples would really help here.

aaschmid commented:
Just as quick test I have removed the 'transitivesness', created a greadle installation, added the dependency to 'prj1' to the example from the ticket, eclipsified it and exported it. I got both 'stax-api' version in the war as the 'prj1' deployment assembly contains it. After manually removing it, it is also gone in the war which means Eclipse resolves dependencies transitively while exporting wars / deploying servers ...

A running example would be great, maybe we can built it upon the test data of ```org.gradle.plugins.ide.eclipse.EclipseIntegrationTest.canCreateAndDeleteMetaData``` which already contains some 'provided' configuration for utility projects ...
(Unfortunately my customers use case and code from a year ago is no longer accessible by me ...)

oehme commented:
We discussed this change a little more internally and concluded that we are not comfortable without some real world projects. We don't know whether the test cases assert the right thing without seeing the change in a more realistic context. I guess you agree given the confusion we had in our previous messages :)

For that reason, I will close this PR for now, to make it clear that we are currently not acting on it.

If you or @ajoberstar can find some real world projects and still want to give this a stab, we would be very happy to support you. Otherwise, we might have an upcoming customer project that will involve fixing this. Although I do not have a timeline for that work yet.

aaschmid commented:
This is of course a big pity because it causes a lot of pain using Eclipse and Gradle in a multi-module context. Not even the easiest thing to get the sources for your dependencies work correctly if the dependencies are contained within the wst-component files ...
:-(

oehme commented:
We recognize that this is a problem, but without realistic examples it is hard to verify. We don't do WTP projects in our team, so we rely on the community to provide the necessary examples.
